flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java

```diff
@@ -428,6 +423,23 @@ public void notifyBufferDestroyed() {
 		// Nothing to do actually.
 	}
 
+	@Override
+	public void onCheckpointBarrier(CheckpointBarrier barrier) {
```

Comment:
Make the following condition out of `synchronized` part, then we do not need to touch the lock for most of the cases.
```
if (!(initialCredit == 0 && options.isExactlyOnceMode() && !options.isUnalignedCheckpoint()))  {
     return;
}
```

Reply:
It might be better to trigger this action by netty thread when received this barrier immediately. Otherwise the task processing might delay much time to better reuse the floating buffers for other channels.

Reply:
Considering the race condition between this operation and canceler task, it might involve in potential conflicts and deadlock case if `removeBufferListener` and adjust `isWaitingForFloatingBuffers` here. Let me think whether we can lazy remove the listener and adjust `isWaitingForFloatingBuffers` by reusing the existing process `RemoteInputChannel#notifyBufferAvailable`.